chips/imxrt10xx/src/gpio.rs

```diff
@@ -436,193 +91,213 @@ enum_from_primitive! {
     }
-
-    pub fn get_port(&self) -> &Port {
-        let port_num: GpioPort = self.get_port_number();
-
-        match port_num {
-            GpioPort::GPIO1 => unsafe { &PORT[0] },
-            GpioPort::GPIO2 => unsafe { &PORT[1] },
-            GpioPort::GPIO3 => unsafe { &PORT[2] },
-            GpioPort::GPIO4 => unsafe { &PORT[3] },
-            GpioPort::GPIO5 => unsafe { &PORT[4] },
-        }
-    }
-
-    // extract the last 6 bits. [6:0] is the pin number, [9:7] is the pad
-    // number
-    pub fn get_pin_number(&self) -> u8 {
-        let mut pin_num = *self as u8;
-
-        pin_num = pin_num & 0b00111111;
-        pin_num
+    /// Returns the pin offset, half-closed range [0, 32)
+    const fn offset(self) -> usize {
+        (self as usize) & 0x3F
     }
 }
 
-enum_from_primitive! {
-    #[repr(u32)]
-    #[derive(PartialEq)]
-
-    /// GPIO pin mode
-    /// In order to set alternate functions such as LPI2C or LPUART,
-    /// you will need to use iomuxc enable_sw_mux_ctl_pad_gpio with
-    /// the specific MUX_MODE according to the reference manual (Chapter 11).
-    /// For the gpio mode, input or output we set the GDIR pin accordingly [^1]
-    ///
-    /// [^1]: 12.4.3. GPIO Programming, page 1008 of the Reference Manual
-    pub enum Mode {
-        Input = 0b00,
-        Output = 0b01
-    }
+/// GPIO pin mode
+///
+/// This describes the pin direction when it's a _GPIO pin_.
+/// It does not describe the direction for other I/O, like LPI2C
+/// or LPUART.
+///
+/// In order to set alternate functions such as LPI2C or LPUART,
+/// you will need to use iomuxc enable_sw_mux_ctl_pad_gpio with
+/// the specific MUX_MODE according to the reference manual (Chapter 11).
+/// For the gpio mode, input or output we set the GDIR pin accordingly [^1]
+///
+/// [^1]: 12.4.3. GPIO Programming, page 1008 of the Reference Manual
+pub enum Mode {
+    Input = 0b00,
+    Output = 0b01,
 }
 
-pub struct Port {
+/// A GPIO port, like `GPIO3`
+///
+/// `Port`s contain collections of pins. Use `Port`s to access pin by their
+/// GPIO offset. See the module-level docs for an example.
+pub struct Port<'a, const N: usize> {
     registers: StaticRef<GpioRegisters>,
-    clock: PortClock,
+    clock: PortClock<'a>,
+    pins: [Pin<'a>; N],
 }
 
-pub static mut PORT: [Port; 5] = [
-    Port {
-        registers: GPIO1_BASE,
-        clock: PortClock(ccm::PeripheralClock::CCGR1(ccm::HCLK1::GPIO1)),
-    },
-    Port {
-        registers: GPIO2_BASE,
-        clock: PortClock(ccm::PeripheralClock::CCGR1(ccm::HCLK1::GPIO1)),
-    },
-    Port {
-        registers: GPIO3_BASE,
-        clock: PortClock(ccm::PeripheralClock::CCGR1(ccm::HCLK1::GPIO1)),
-    },
-    Port {
-        registers: GPIO4_BASE,
-        clock: PortClock(ccm::PeripheralClock::CCGR1(ccm::HCLK1::GPIO1)),
-    },
-    Port {
-        registers: GPIO5_BASE,
-        clock: PortClock(ccm::PeripheralClock::CCGR1(ccm::HCLK1::GPIO1)),
```

Comment:
If reverting to this change, update the `PeripheralClock`s. Before this PR, GPIO2 through 5 would control GPIO1's clock gate.